### PR TITLE
Evolve task weights

### PR DESCRIPTION
**Pull request by srichers**

### CODE REVIEW

File: model_training/ml_trainmodel.py [hunk 1]
```diff
@@ -236,7 +236,7 @@ def accumulate_stable_loss(dataset_list, traintest):
         print(f"{epoch:4d}  {loss_dict['learning_rate']:12.5e}  {loss_dict['train_loss']:12.5e}  {loss_dict['test_loss']:12.5e}")
         if(epoch%parms["output_every"]==0):
             outfilename = os.getcwd()+"/model"+str(epoch)
-            F4i = dataset_asymptotic_test_list[0].datasets[0]
+            F4i = dataset_asymptotic_test_list[0].tensors[0]
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py:258]: this seems reasonable

**Review by jmcguigan10 — APPROVED** [review]

File: model_training/ml_trainmodel.py [hunk 2]
```diff
@@ -136,21 +133,26 @@ def contribute_loss(pred, true, traintest, key, loss_fn):
             ndens_pred_train, fluxmag_pred_train, Fhat_pred_train = get_ndens_fluxmag_fhat(F4f_pred_train)
             ndens_true_train, fluxmag_true_train, Fhat_true_train = get_ndens_fluxmag_fhat(F4f_true_train)
 
+            # reset the loss and gradients
+            batch_loss = torch.tensor(0.0, requires_grad=True)
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py]: This doesn't specify a device which could be an issue. I would change it to something like ` batch_loss = torch.tensor(0,0, device= parms["device"])` or use the torch zeroes method:   ` batch_loss = torch.zeros((), device=parms["device"])`. I don't think you you need to set `requires_grad=True` here since it shouldn't be a cpu scalar.

File: model_training/ml_trainmodel.py [hunk 3]
```diff
@@ -136,21 +133,26 @@ def contribute_loss(pred, true, traintest, key, loss_fn):
...
-            batch_loss = batch_loss +  direction_loss_fn(Fhat_pred_train, Fhat_true_train)
-            batch_loss = batch_loss + comparison_loss_fn(growthrate_pred_train/ntotal(F4i_asymptotic_train), torch.log(growthrate_true_train/ntotal(F4i_asymptotic_train)/ndens_to_invsec))
+            batch_loss = batch_loss + torch.exp(-model.log_task_weights["stability"] ) * stability_loss_fn(stable_pred_train, stable_true_train)
+            batch_loss = batch_loss + torch.exp(-model.log_task_weights["ndens"]     ) * comparison_loss_fn(ndens_pred_train, ndens_true_train)
+            batch_loss = batch_loss + torch.exp(-model.log_task_weights["fluxmag"]   ) * comparison_loss_fn(fluxmag_pred_train, fluxmag_true_train)
+            batch_loss = batch_loss + torch.exp(-model.log_task_weights["direction"] ) *  direction_loss_fn(Fhat_pred_train, Fhat_true_train)
+            batch_loss = batch_loss + torch.exp(-model.log_task_weights["growthrate"]) * comparison_loss_fn(growthrate_pred_train/ntotal(F4i_asymptotic_train), torch.log(growthrate_true_train/ntotal(F4i_asymptotic_train)/ndens_to_invsec))
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py]: It kinda seems like we're comparing apples to oranges here by comparing linear to log. Also I'm not sure why you divide one by `ndens_to_invsec` but not the other.

File: model_training/ml_trainmodel.py [hunk 4]
```diff
@@ -161,21 +163,21 @@ def contribute_loss(pred, true, traintest, key, loss_fn):
...
-                total_loss = total_loss + contribute_loss(Fhat_pred, Fhat_true, traintest, "direction", direction_loss_fn)
-                total_loss = total_loss + contribute_loss(growthrate_pred/ntotal(F4i), torch.log(growthrate_true/ntotal(F4i)/ndens_to_invsec), traintest, "growthrate", comparison_loss_fn)
+                total_loss = total_loss + torch.exp(-model.log_task_weights["ndens"]     ) * contribute_loss(ndens_pred, ndens_true, traintest, "ndens", comparison_loss_fn)
+                total_loss = total_loss + torch.exp(-model.log_task_weights["fluxmag"]   ) * contribute_loss(fluxmag_pred, fluxmag_true, traintest, "fluxmag", comparison_loss_fn)
+                total_loss = total_loss + torch.exp(-model.log_task_weights["direction"] ) * contribute_loss(Fhat_pred, Fhat_true, traintest, "direction", direction_loss_fn)
+                total_loss = total_loss + torch.exp(-model.log_task_weights["growthrate"]) * contribute_loss(growthrate_pred/ntotal(F4i), torch.log(growthrate_true/ntotal(F4i)/ndens_to_invsec), traintest, "growthrate", comparison_loss_fn)
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py]: Same here— you take the log of one but not the other. Might be an issue.

File: model_training/ml_neuralnet.py [hunk 5]
```diff
@@ -29,6 +29,12 @@ def __init__(self, parms):
         # store input arguments
         self.NF = parms["NF"]
 
+        # store loss weight parameters for tasks
+        self.log_task_weights = nn.ParameterDict({
+            name: nn.Parameter(torch.tensor(parms[f"log_task_weight_{name}"], dtype=torch.float32))
+            for name in ["stability", "growthrate", "ndens", "fluxmag", "direction", "unphysical"]
```
Comment (jmcguigan10) [inline model_training/ml_neuralnet.py:35]: Here you're telling the model that it is allowed to learn the unphysical weight, which I don't think we want.

File: model_training/ml_trainmodel.py [hunk 6]
```diff
@@ -136,21 +133,26 @@ def contribute_loss(pred, true, traintest, key, loss_fn):
...
+            batch_loss = batch_loss + torch.exp(-model.log_task_weights["stability"] ) * stability_loss_fn(stable_pred_train, stable_true_train)
+            batch_loss = batch_loss + torch.exp(-model.log_task_weights["ndens"]     ) * comparison_loss_fn(ndens_pred_train, ndens_true_train)
+            batch_loss = batch_loss + torch.exp(-model.log_task_weights["fluxmag"]   ) * comparison_loss_fn(fluxmag_pred_train, fluxmag_true_train)
+            batch_loss = batch_loss + torch.exp(-model.log_task_weights["direction"] ) *  direction_loss_fn(Fhat_pred_train, Fhat_true_train)
+            batch_loss = batch_loss + torch.exp(-model.log_task_weights["growthrate"]) * comparison_loss_fn(growthrate_pred_train/ntotal(F4i_asymptotic_train), torch.log(growthrate_true_train/ntotal(F4i_asymptotic_train)/ndens_to_invsec))
             if parms["do_unphysical_check"]:
-                batch_loss = batch_loss + 100 * unphysical_loss_fn(F4f_pred_train/ntotal(F4i_asymptotic_train)[:,None,None,None], None)
+                batch_loss = batch_loss + torch.exp(-model.log_task_weights["unphysical"]) * unphysical_loss_fn(F4f_pred_train/ntotal(F4i_asymptotic_train)[:,None,None,None], None)
```
Comment (jmcguigan10) [inline model_training/ml_trainmodel.py:150]: Like I said above, I don't think we want to tell the model to learn that it can do "unphysical" things. We should change this back to: `batch_loss = batch_loss + parms["lambda_unphysical"] * unphysical_loss_fn(...)`

**Review by jmcguigan10 — COMMENTED** [review]
Looks pretty good, I just have a few comments.

File: model_training/ml_trainmodel.py (hunk 2 above)
Comment (srichers) [reply model_training/ml_trainmodel.py]: Usually it throws an error if tensors on different devices are used together. But it's a good point - I'll make the change.

File: model_training/ml_trainmodel.py (hunk 6 above)
Comment (srichers) [reply model_training/ml_trainmodel.py:150]: I think it might be worth seeing how well it learns to not do unphysical things. We have a means of treating those cases, since there will always be times when an unphysical result is predicted. And I would like it to not prevent other things from being learned. Of course, if the unphysical errors become large it would make sense to keep the weight large.

File: model_training/ml_pytorch.py [hunk 7]
```diff
@@ -54,32 +55,32 @@
...
     parms["amsgrad"]= False
-    parms["weight_decay"]= 1e-2 #1e-5
+    parms["weight_decay"]= 0 #1e-2 #1e-5
     parms["learning_rate"]= 2e-4 # 1e-3
     parms["fused"]= True
     parms["patience"]= 500
     parms["cooldown"]= 500
     parms["factor"]= 0.5
-    parms["warmup_iters"]=10
+    parms["warmup_iters"]=0
```
Comment (jmcguigan10) [inline model_training/ml_pytorch.py:83]: I think we should keep the warmup.

**Review by jmcguigan10 — COMMENTED** [review]
Looks good.

**Review by jmcguigan10 — APPROVED** [review]
